lib/puppet/context.rb

```diff
@@ -103,14 +81,63 @@ def mark(name)
   #
   # @api private
   def rollback(name)
-    if @rollbacks[name].nil?
+    if @rollbacks.value[name].nil?
       raise UnknownRollbackMarkError, _("Unknown mark '%{name}'") % { name: name }
     end
 
-    while @stack[-1][0] != @rollbacks[name]
-      pop
+    @stack.value = @rollbacks.value[name]
+    rollbacks_without_name = @rollbacks.value.dup
+    rollbacks_without_name.delete(name)
+    @rollbacks.value = rollbacks_without_name
+  end
+
+  class Puppet::Context::EmptyStack
+    def lookup(name, &block)
+      if block
+        block.call
+      else
+        raise UndefinedBindingError, _("Unable to lookup '%{name}'") % { name: name }
+      end
+    end
+
+    def pop
+      raise(StackUnderflow, _("Attempted to pop, but already at root of the context stack."))
+    end
+
+    def push(overrides)
+      Puppet::Context::Stack.new(self, overrides)
+    end
+
+    def bindings
+      {}
     end
+  end
+
+  class Puppet::Context::Stack
+    attr_reader :bindings
 
-    @rollbacks.delete(name)
+    def initialize(parent, bindings)
+      @parent = parent
+      @bindings = parent.bindings.merge(bindings || {})
+    end
+
+    def lookup(name, &block)
+      if @bindings.include?(name)
+        value = @bindings[name]
+        value.is_a?(Proc) ? (@bindings[name] = value.call) : value
```

Comment:
The above means two lookups for the most common case of finding the value. It is faster to lookup, and if value is nil, check if this means "not found or bound nil value" by testing with `include?`

Reply:
This part is unchanged, just moved around. Unless it's showed up as a hot spot that seems like an unnecessary micro optimization.